### PR TITLE
Store bindings

### PR DESCRIPTION
**Pull request by pavelmaliy**
# Pull Request Template

## Prerequisites

* If there is no issue related to your pull request - open one and assign yourself to it. If you're proposing a solution to an already opened issue - simply assign yourself to it.

## Motivation

Store bindings in order to be able to delete orphan instances in case platform was terminated

## Approach

if bindings are stored SM can safely do unbind and finally delete the orphan instances

## Pull Request status

* [x] Initial implementation
* [x] Refactoring
* [x] Unit tests
* [x] Integration tests

Feel free to construct the checklist with whatever items seem most reasonable for your change. You could disassemble the Implementation part to even smaller separate checklist items if you're working on something big for example. But do make the effort to provide a checklist of some sort so that the core team as well as the community can have an idea about the progress of your Pull Request.

## Third-party code

If you use third party code with your contribution such as, components, libraries, or snippets make sure to mention that.

Make sure that your contribution complies with our [Contribution guidelines][1].

## Work in Progress label

For as long as development of your Pull Request is still ongoing you MUST label it with a **wip** label as well as prefix the name of the PR with *[WIP]*.

For example: *[WIP] Service brokers API*

[1]: https://github.com/Peripli/service-manager/blob/master/CONTRIBUTING.md

### CODE REVIEW

File: api/osb/osb_store_plugin.go [hunk 1]
```diff
@@ -0,0 +1,1113 @@
...
+type entityOperation string
+
+const (
+	READY    entityOperation = "ready"
+	ROLLBACK entityOperation = "rollback"
+	DELETE   entityOperation = "failed"
+	NONE     entityOperation = "none"
+)
+
+type BindingStorageKey struct{}
```
Comment (dotchev) [inline api/osb/osb_store_plugin.go:61]: To ensure encapsulation context keys should use a private type and provide strongly-typed access functions to specific values. See pkg/web/context_keys.go

Comment (dotchev) [reply api/osb/osb_store_plugin.go]: Also add tests for this context value
You can do that by registering a filter that sets this context value depending on some criteria, e.g. global var. You can register this filter in test/osb_test/osb_suite_test.go in `WithSMExtensions`.

Comment (pavelmaliy) [reply api/osb/osb_store_plugin.go]: done

File: api/osb/osb_store_plugin.go [hunk 2]
```diff
@@ -0,0 +1,1102 @@
...
+	}
+
+	if _, err := storage.Update(ctx, serviceInstance, types.LabelChanges{}); err != nil {
+		return util.HandleStorageError(err, string(serviceInstance.GetType()))
+	}
+
+	return nil
+}
+
+func (sp *StorePlugin) updateEntityReady(ctx context.Context, storage storage.Repository, bindingID string, objectType types.ObjectType) error {
```
Comment (dotchev) [inline api/osb/osb_store_plugin.go:926]: `bindingID` -> `resourceID`

File: api/osb/osb_store_plugin.go [hunk 3]
```diff
@@ -0,0 +1,1102 @@
...
+func (sp *StorePlugin) createOsbEntity(status int, storeOp func(state types.OperationState, category types.OperationCategory) error, storeEntity func(ready bool) error) error {
+	switch status {
+	case http.StatusCreated:
+		if err := storeOp(types.SUCCEEDED, types.CREATE); err != nil {
+			return err
+		}
+		if err := storeEntity(true); err != nil {
+			return err
+		}
+	case http.StatusOK:
```
[the remark this answers is not in the document]
Comment (pavelmaliy) [reply api/osb/osb_store_plugin.go:1085]: they're not on status OK I don't want to store op always

File: api/osb/osb_store_plugin.go [hunk 4]
```diff
@@ -0,0 +1,1102 @@
...
+// NewStorePlugin creates a plugin that stores service instances on OSB requests
+func NewStorePlugin(repository storage.TransactionalRepository) *StorePlugin {
+	return &StorePlugin{
+		Repository: repository,
+	}
+}
+
+// StoreServiceInstancePlugin represents a plugin that stores service instances and bindings on OSB requests
+type StorePlugin struct {
+	Repository storage.TransactionalRepository
```
[the remark this answers is not in the document]
Comment (dotchev) [reply api/osb/osb_store_plugin.go:289]: `Repository` field should be private, `StorePlugin` struct should be public as it is returned by the constructor

**Pull request merged**